### PR TITLE
Document precondition of SourceFile's constructor

### PR DESCRIPTION
**Pull request by tothambrus11**
I added documentation for the precondition mentioned in: #1489 
When given the same `baseName` for a SourceFile's constructor as before, the provided file content must equal the one before for synthesized source files.

I added a runtime check that only runs in debug builds, so it shouldn't influence the performance of the compiler in release mode.

### CODE REVIEW

File: Sources/FrontEnd/SourceFile.swift [hunk 1]
```diff
@@ -61,9 +61,19 @@ public struct SourceFile {
   }
 
   /// Creates a synthetic source file with the specified contents and base name.
+  ///
+  /// If `baseName` is not provided, a unique name will be generated.
+  ///
+  /// - Precondition: the same baseName must not be used for different text, as
+  ///   synthesized files with the same `baseName` will share storage.
+  ///
```
Comment (kyouko-taiga) [inline Sources/FrontEnd/SourceFile.swift:69]:
```suggestion
  /// - Precondition: No synthetic source file named `baseName` has been created yet.
```

File: Sources/FrontEnd/SourceFile.swift [hunk 2]
```diff
@@ -61,9 +61,19 @@ public struct SourceFile {
...
   public init(synthesizedText text: String, named baseName: String = UUID().uuidString) {
     let storage = Storage(URL(string: "synthesized://\(baseName)")!) { text[...] }
     self.storage = storage
+
+    // The assertion is not evaluated in release builds.
+    assert(text == self.storage.text, "Precondition violated: synthesized files with the " +
+      "same baseName must have the same text globally.")
```
Comment (kyouko-taiga) [inline Sources/FrontEnd/SourceFile.swift:76]:
```suggestion
    assert(text == self.storage.text, "duplicate synthesized filename")
```

**Comment by tothambrus11** [comment]
Thanks for the rewording, I applied the changes.

**Comment by dabrahams** [comment]
I'm confused. Didn't we say in Slack that this isn't the right fix? Don't we just want to eliminate the parameter at all allows users to bypass the UUID generation for synthesized files?

**Comment by tothambrus11** [comment]
You are right, I was confused about the conclusion. @kyouko-taiga can you clarify what should be the solution?

**Comment by kyouko-taiga** [comment]
The reason why the default value of the parameter can be overridden is that mangling tests rely on the source file having a specific name. If we remove this parameter we'll also have to change the way mangling is tested.

**Comment by kyouko-taiga** [comment]
The agreed-upon fix has been implemented by #1530.

**Pull request closed without merging**